### DISCUSSION TITLE
how to convert 360 fisheye to rectangular

### DISCUSSION THREAD

**Discussion by thijs1970** [ImageMagick/ImageMagick#5316] [Help] [answered]
I'm trying to convert an image from a 360 degrees camera to a flat rectangular image. But I don't get it to work. I tried barrel conversions, perspective and so on but nothing gives the desired result. I expect it to be very simple.

A basic example is how to convert input.jpg to output.jpg?
![input](https://user-images.githubusercontent.com/88130831/178322891-d6b79397-9e6f-4c6e-b302-aa2ad333bfa5.jpg) ![output](https://user-images.githubusercontent.com/88130831/178322916-28c6c921-24e7-4831-a633-2d923e44b985.jpg)

**Comment by fmw42** [comment]
It is a conformal mapping.  See

https://arxiv.org/pdf/1509.06344.pdf
http://squircular.blogspot.com/2015/09/mapping-circle-to-square.html

**Reply by thijs1970** [reply]
Hello Fred,
Thanks for your reply. Is there an imagemagick command to do this conversion?

**Reply by fmw42** [reply]
Create the equations as displacement maps. Then use -compose displace -composite to apply them. As I have time, I will try to add this to my Imagemagick conformal script or create a new one for it. I have two Python/OpenCV scripts already that do that, if that will help you.

**Comment by fmw42** [comment, marked as answer]
Here is one way below. But note that is not specific for a proper fisheye. If you want a proper conversion see my fisheye2rect and defisheye scripts, for example, at http://www.fmwconcepts.com/imagemagick

Input:

![red_rings](https://user-images.githubusercontent.com/13458218/178373505-e4f6591e-dc5a-4bed-91e0-ef92e7d0d3f6.jpg)

Unix syntax:

```
magick red_rings.jpg -fx \
"xc=w/2; yc=h/2; ii=(i-xc)/xc; jj=(j-yc)/yc; \
iii=xc*ii*sqrt(1-0.5*jj^2)+xc; jjj=yc*jj*sqrt(1-0.5*ii^2)+yc; \
u.p{iii,jjj}" x.png
```

![x](https://user-images.githubusercontent.com/13458218/178373553-86157259-f383-43de-b30a-303bda8c852d.png)

**Reply by thijs1970** [reply]
Thanks. This command looks very complex. And unfortunately it's also much slower than the tool I'm using now (ffmpeg).  I was hoping magick had a simple and fast command for doing my conversion.

The ffmpeg command I'm using is:
ffmpeg -i input.jpg -vf v360=input=fisheye:output=equirect:ih_fov=190:iv_fov=190:h_fov=180:v_fov=180 output.jpg

So I probably stick to ffmpeg for doing my conversions.

Thanks anyway for your help in answering my request.

**Comment by snibgo** [comment]
> And unfortunately it's also much slower 

What version of IM do you use? Until about v7.1.0-21, "-fx" was slow. Now it is much faster.

**Reply by thijs1970** [reply]
I have installed version "7.1.0-43" now and when testing ffmpeg is around 5 times faster than IM with the -fx option. I have to do thousands of conversions so speed is important for me.   I use IM for rotating, resizing, gamma correction and so. It's perfect for that, but for the fisheye conversion I must make use of ffmpeg.

**Comment by fmw42** [comment]
You can try my script defisheye. But it uses -fx and will need a current version of IM 7 to process it fast. However, I suspect that ffmpeg is highly optimized.

**Comment by thijs1970** [comment]
FFMPEG is 5-6 times faster than IM with the -fx option, so ffmpeg is the tool to use for me for the fisheye conversions.

**Reply by fmw42** [reply]
As I said, ffmpeg is optimized. It is compiled code.  -fx expressions are not compiled, but has been improved vastly.

**Comment by snibgo** [comment]
> FFMPEG is 5-6 times faster than IM with the -fx option, ...

Please say what version of IM you use.

**Reply by thijs1970** [reply]
If have installed the latest version "7.1.0-43".

**Comment by snibgo** [comment]
Okay. IM doesn't contain compiled code to rectify fisheye distortions, and "-fx" will in general be slower than compiled code.
